### PR TITLE
support running envoy on specific nodes

### PR DESCRIPTION
**Pull request by amdonov**
Trying to gauge willingness to accept this feature. PR is not ready to merge as is. Needs more comments and tests. We like to run envoy pods on a dedicated node pool. This would allow use of node selectors and tolerations to target specific nodes.

### CODE REVIEW

File: api/v1alpha1/contour_types.go [hunk 1]
```diff
@@ -229,6 +229,41 @@ type EnvoyNetworkPublishing struct {
...
 	ContainerPorts []ContainerPort `json:"containerPorts,omitempty"`
+
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// +optional
+	Tolerations []Toleration `json:"tolerations,omitempty"`
+}
+
+type Toleration struct {
```
Comment (skriss) [inline api/v1alpha1/contour_types.go:240]: any particular reason to use a custom type here, vs. reusing the upstream `corev1.Toleration`?

File: api/v1alpha1/contour_types.go [hunk 2]
```diff
@@ -229,6 +229,41 @@ type EnvoyNetworkPublishing struct {
 	// +kubebuilder:validation:MaxItems=2
 	// +kubebuilder:default={{name: http, portNumber: 8080}, {name: https, portNumber: 8443}}
 	ContainerPorts []ContainerPort `json:"containerPorts,omitempty"`
+
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// +optional
+	Tolerations []Toleration `json:"tolerations,omitempty"`
```
Comment (skriss) [inline api/v1alpha1/contour_types.go:237]: I'm not sure these belong under `NetworkPublishing`, we may want to move them elsewhere in the API.

**Review by skriss — COMMENTED** [review]
Thanks for the PR @amdonov! I definitely think this feature makes sense to add. xref #362 and cc @danehans for thoughts as well

File: api/v1alpha1/contour_types.go (hunk 1 above)
Comment (danehans) [reply api/v1alpha1/contour_types.go:240]: I prefer using `corev1.Toleration` instead of supporting a custom type.

**Comment by amdonov** [comment]
Sounds good. I'll close this PR and open two different ones following your
recommendations.

**Comment by amdonov** [comment]
Made custom types for ContourNodePlacement and EnvoyNodePlacement as requested. Can't think of any fields that would differ between the two. Recommend using a single type if you agree.

**Comment by danehans** [comment]
@amdonov I started providing feedback, but I stopped since I felt it was becoming confusing. Instead, I created a [commit](https://github.com/danehans/contour-operator/commit/c383af4d5429a6781b867eac2e163b6ffb05478f) that details the suggested changes, PTAL when you have a moment. Tests are still needed and I want to make sure you and others agree to use `node-role.kubernetes.io/worker: ""` as the default node selector.

> Made custom types for ContourNodePlacement and EnvoyNodePlacement as requested. Can't think of any fields that would differ between the two. Recommend using a single type if you agree.

Although the two types are the same today, they may differ in the future. Therefore I prefer defining separate types.

cc: @stevesloka @skriss @sunjayBhatia @youngnick

**Comment by amdonov** [comment]
@danehans I'd recommend not setting a default node selector. I checked labels on a couple of the k8s distros that I use frequently. For the control plane nodes one had node-role.kubernetes.io/master=true, and the other had node-role.kubernetes.io/master="". However, the worker nodes weren't labeled in either case. I'm afraid the pods wouldn't get scheduled and could lead to confusion.

**Comment by danehans** [comment]
> I'd recommend not setting a default node selector.

@amdonov thanks for looking into this. I was unsure how distros handled the node roles by default. With that said, here is the [updated commit](https://github.com/danehans/contour-operator/commit/b118f5459c11bb2c66bceaaf836ce7be7e56f295). Again, unit tests are still required.

**Comment by amdonov** [comment]
I'm done pending further feedback.

File: examples/operator/operator.yaml [hunk 3]
```diff
@@ -403,6 +403,30 @@ spec:
                               information specific to the underlying infrastructure
                               provider.
                             properties:
+                              aws:
```
Comment (danehans) [inline examples/operator/operator.yaml:406]: __Note__: This is leftover from https://github.com/projectcontour/contour-operator/pull/336

**Review by danehans — APPROVED** [review]

File: api/v1alpha1/helpers.go [hunk 4]
```diff
@@ -51,3 +51,47 @@ func (c *Contour) IsFinalized() bool {
...
+	if c.Spec.NodePlacement != nil &&
+		c.Spec.NodePlacement.Contour != nil &&
+		len(c.Spec.NodePlacement.Contour.Tolerations) > 0 {
+		return true
+	}
+
+	return false
+}
+
+// EnvoyNodeSelectorExists returns true if a nodeSelector is specified for Contour.
```
Comment (skriss) [inline api/v1alpha1/helpers.go:77]:
```suggestion
// EnvoyNodeSelectorExists returns true if a nodeSelector is specified for Envoy.
```

File: api/v1alpha1/helpers.go [hunk 5]
```diff
@@ -51,3 +51,47 @@ func (c *Contour) IsFinalized() bool {
...
+	if c.Spec.NodePlacement != nil &&
+		c.Spec.NodePlacement.Envoy != nil &&
+		c.Spec.NodePlacement.Envoy.NodeSelector != nil {
+		return true
+	}
+
+	return false
+}
+
+// EnvoyTolerationsExist returns true if a nodeSelector is specified for Envoy.
```
Comment (skriss) [inline api/v1alpha1/helpers.go:88]:
```suggestion
// EnvoyTolerationsExist returns true if tolerations are set for Envoy.
```

**Review by skriss — CHANGES_REQUESTED** [review]
couple godoc nits but otherwise LGTM, thanks @amdonov!

**Review by skriss — APPROVED** [review]
LGTM. DCO bot is unhappy with the GH suggestion commits but we can address that when we squash+merge.

**Pull request merged**